src/Redistribution.sol

```diff
@@ -534,56 +551,57 @@ contract Redistribution is AccessControl, Pausable {
-
-            // question is whether randomNumber / MaxH < probability
-            // where probability is stakeDensity / currentSum
-            // to avoid resorting to floating points all divisions should be
-            // simplified with multiplying both sides (as long as divisor > 0)
-            // randomNumber / (MaxH + 1) < stakeDensity / currentSum
-            // ( randomNumber / (MaxH + 1) ) * currentSum < stakeDensity
-            // randomNumber * currentSum < stakeDensity * (MaxH + 1)
-            if (randomNumberTrunc * currentSum < currentReveals[i].stakeDensity * (uint256(MaxH) + 1)) {
-                truthRevealedHash = currentReveals[i].hash;
-                truthRevealedDepth = currentReveals[i].depth;
+                if (randomNumberTrunc * currentSum < currentReveals[revIndex].stakeDensity * (uint256(MaxH) + 1)) {
+                    truthRevealedHash = currentReveals[revIndex].hash;
+                    truthRevealedDepth = currentReveals[revIndex].depth;
+                }
+
+                k++;
             }
         }
 
         emit TruthSelected(truthRevealedHash, truthRevealedDepth);
 
-        string memory winnerSelectionAnchor = currentWinnerSelectionAnchor();
-
-        uint256 k = 0;
-        for (uint256 i = 0; i < revealsArrayLength; i++) {
-            if (truthRevealedHash == currentReveals[i].hash && truthRevealedDepth == currentReveals[i].depth) {
-                currentWinnerSelectionSum += currentReveals[i].stakeDensity;
-                randomNumber = keccak256(abi.encodePacked(winnerSelectionAnchor, k));
+        k = 0;
 
-                randomNumberTrunc = uint256(randomNumber & MaxH);
+        string memory winnerSelectionAnchor = currentWinnerSelectionAnchor();
 
-                if (
-                    randomNumberTrunc * currentWinnerSelectionSum < currentReveals[i].stakeDensity * (uint256(MaxH) + 1)
-                ) {
-                    winner = currentReveals[i];
+        for (uint256 i = 0; i < commitsArrayLength; i++) {
+            revIndex = currentCommits[i].revealIndex;
+            if (currentCommits[i].revealed ) {
+               if ( truthRevealedHash == currentReveals[revIndex].hash && truthRevealedDepth == currentReveals[revIndex].depth) {
+                    currentWinnerSelectionSum += currentReveals[revIndex].stakeDensity;
+                    randomNumber = keccak256(abi.encodePacked(winnerSelectionAnchor, k));
+    
+                    randomNumberTrunc = uint256(randomNumber & MaxH);
+    
+                    if (
+                        randomNumberTrunc * currentWinnerSelectionSum < currentReveals[revIndex].stakeDensity * (uint256(MaxH) + 1)
+                    ) {
+                        winner = currentReveals[revIndex];
+                    }
+    
+                    k++;
+                } else {
+                    Stakes.freezeDeposit(currentReveals[revIndex].overlay, 3 * roundLength * uint256(2**truthRevealedDepth));
```

Comment:
shall we make a constant for the not_reveal_penalty_rounds and disagreement_frozen_penalty_rounds or something similar?